### ISSUE TITLE
Add option to keep license for certain tree-shaken dependencies

### ISSUE THREAD

**Issue by timmb** [mjeanroy/rollup-plugin-license#1092]
Since https://github.com/mjeanroy/rollup-plugin-license/issues/380, the plugin no longer includes licenses for dependencies that have been removed through tree shaking. However, there are occasions where you want to keep them in.

In my case, I am using GLSL code from node modules which is being preprocessed and imported by glslify. The dependency is shaken out but the 3rd party code remains imported within my own GLSL code.

In this situation, it would be helpful to add an option to manually force certain node modules to have their licenses included. 

For example,

```
thirdParty: {
  includePrivate: true,
  output: `${dest}/dependencies.txt`,
  alwaysInclude: [
    'glsl-easings',
  ]     
},
```

**Comment by mjeanroy** [comment]
Hi @timmb,

Thanks for the issue!
I'm not sure to understand exactly the problem, could you share a test repo where I can reproduce first?
Thanks!

**Comment by timmb** [comment]
Hi @mjeanroy 

Please see here: https://github.com/timmb/mwe-rollup-plugin-license-missing-glsl-licenses

Run `npm run build`

Code from `glsl-noise` is bundled into the output via the `glslify` preprocessor, but this is invisible to `rollup-plugin-license` and the license is not included in `dependencies.txt`

I don't anticipate this plugin working with glslify but I currently have no option to make it include the licence for `glsl-noise`.

See the README in the demo repo for more detail.

**Comment by mjeanroy** [comment]
Hi @timmb,

I'm a bit torn with your proposal: in the example you shared above, `glslify` is only a build dependency used to compile your own code, it's not used in your final bundle.

Furthermore, `glslify` should not be a direct dependency of your package, since you don't need it at runtime, it's a `devDependency`.

I'm not sure if including build dependency in the dependency file output is the right thing to do.
What would be the reason for that?

Btw, a bit out of scope, but #380 is not the cause of your issue, as `glslify` is not bundled at all.

**Comment by mjeanroy** [comment]
Thanks @timmb for all the details, I definitely agree with your conclusion:

> As code from glsl-noise is in the final distribution, its license also needs to be included.

Let me dig a bit deeper, I would like to check if this issue can be fixed without adding a new option.

**Comment by mjeanroy** [comment]
Hi @timmb,

The main problem here is that the glslify rollup plugin transform the code, and rollup has no way to know which modules have been loaded or not (another side effect is that files are not watchable by rollup as well).

I updated the plugin implementation to make rollup aware of these files, you can see [here](https://github.com/mjeanroy/mwe-rollup-plugin-license-missing-glsl-licenses/tree/feat/emit-static-files).

You can review the diff commit by commit, the [last commit](https://github.com/timmb/mwe-rollup-plugin-license-missing-glsl-licenses/commit/cd625588113baca4bb921cd93a13aa391d9556b7) is where the "magic" happens.

Basically, what I did:
- Use `depper.on` to be aware of loaded files.
- Tweak a bit the output code to import these files (and replace it by a single line of comment).

Then, the file is correctly loaded, and everything works fine.

I don't know if it could be simplified, I would love to see a way to force rollup to include modules in the chunk modules, that could make things way easier.

**Comment by nborko** [comment]
As I think about it, it's not really the same issue. The issue with tailwind is that it gets generated via a postcss plugin, so it's not being directly imported in a way that rollup can see it.

But it would be nice if the license plugin had an option to add package names of packages that exist in node_modules as a dependency that get built into a bundled product, even if rollup doesn't itself see it during the bundling process.

**Comment by mitar** [comment]
For TailwindCSS, I must say that I think (but I am not a lawyer) that one does not have to include a banner for it. You are not bundling their code, just the output of the program. See my comments [here](https://opensource.stackexchange.com/a/14626/29567) and [here](https://github.com/tailwindlabs/tailwindcss/discussions/11094#discussioncomment-8629564).

**Comment by unsungNovelty** [comment]
@mitar - Excuse the delay. I don't use Github. Which is why I missed this.

> For TailwindCSS, I must say that I think (but I am not a lawyer) that one does not have to include a banner for it. You are not bundling their code, just the output of the program.

No. You don't need to include the license of the Tailwind CLI which is generating the output. But you still need to include the TailwindCSS framework's license somewhere. I have put it in https://www.unsungnovelty.org/license/.

The reason why I opened that ticket you have linked is because people (including me) don't know how to comply with the license. And the fact that I was just learning about licenses and found it very interesting is all. Nobody is even gonna know that they are in MIT license violation. If the end output could include the MIT license text, automatically everyone who uses TailwindCSS will be covered. Otherewise, every single person has to have a process to handle this. Thats 1000's or probably 100's of thousands of websites/projects. So I thought if Tailwind could include it, it would make it easier for everyone. But that is not the project's responsbility at all _technically_.

> Creator of TailwindCSS https://github.com/tailwindlabs/tailwindcss/issues/11536#issuecomment-1618990213.

Adam's remarks is with respect to the angle of the Project itself. His side is correct. His project is complying to the usage of MIT license correctly. The users are not. It's the user's (TailwindCSS framework's user's) responsibility to handle this. Which means, people who don't have a process is violating the Tailwind CSS license and is hence void from any of the freedom's MIT is supposed to be giving you.

Since most license works with the benefit of the doubt, it's overlooked. I was just learning about the license and hence raised it out of curiousity is all.

**Comment by unsungNovelty** [comment]
> I am not sure what "TailwindCSS framework" is. In my view, Tailwind CSS CLI is a compiler of CSS and as such output of Tailwind CLI is not licensable at all.

OK. That is why I mentioned, in this case, that you wouldn't have to include the License of the Tailwind CLI itself. But as per MIT license, you are supposed to include the license and copyright notice.

<img width="1280" height="242" alt="Image" src="https://github.com/user-attachments/assets/74610396-0008-4cc0-aaa1-fb9621d06e1f" />

> There is no "creative input or decisions" (except for mine, when selecting classes) for intellectual property to apply to output.

But this the _creative input or decisions_ part. You are literally making the entire layout which includes all sorts of components and elements. You're not choosing from pre-built options like which package manager to use?:

- [ ] NPM
- [ ] Yarn
- [x] PNPM

Which is something you would see when setting up projects with Vite CLI etc.
Also, I am trying to use your with libraries. After tree-shaking etc, why would I have to include license text to any libraries I use then?

My understanding is this is where the below lince comes into play:
 _The above copyright notice and this permission notice shall be included **in all copies or substantial portions** of the Software._

**Comment by mitar** [comment]
> You are literally making the entire layout which includes all sorts of components and elements.

Yes, I am doing it, not authors of tailwind. So if something, I could require how that CSS output is licensed, under which license and that if I pick MIT license, that a copyright notice attributing **me** should be made available. But I could simply license it together with all the code I made for frontend as well, not handling CSS in any special way.

Are you including on frontend license for NPM? node.js itself? postcss? autoprefixer? No. Because those are just tools. Like tailwind CSS (by [author's own comment](https://github.com/tailwindlabs/tailwindcss/issues/11536#issuecomment-1618990213)).

The point you are arguing has been argued for decades on FOSS community about the role of compilers and other tools in a build pipeline and where do you draw a line. Hey, do you have to also include your OS license if you are compiling on open source operating system? All popular OSI licenses (if not all) do not require from you to attribute any of that.

So sure, feel free to include more than you need. But generally, if you do not include creative work of somebody else in your creative work (so if it is tree-shaked out), you do not have to attribute it. Of course, you can always do so because you want to promote it or something.

Anyway, this is getting very off topic. I do think it might be useful for this plugin to have an override mechanism like mentioned in the original comment, where some tooling pulls additional code which is not a direct dependency of the project itself and not seen by rollup. I do think though that this will eventually lead to more problems for you when using rollup because if rollup is not aware of it, then it cannot do also other potential things about that code. Like maybe you just need only part of that additional code and it could be further tree shook. It [looks like there are other compatibility issues](https://github.com/glslify/glslify/issues/131). So maybe glslify has to be rollup plugin as well, or support it directly? Like it does browserify?

**Comment by unsungNovelty** [comment]
> Are you including on frontend license for NPM? node.js itself? postcss? autoprefixer? No. Because those are just tools. Like tailwind CSS (by https://github.com/tailwindlabs/tailwindcss/issues/11536#issuecomment-1618990213).

I thought I clarified the part about Adam's comment. He isn't really bothered about the MIT license of his users. HE IS TALKING FROM THE PROJECT'S POV and **NOT from the end-user's POV** which I am talking about.

> The point you are arguing has been argued for decades on FOSS community about the role of compilers and other tools in a build pipeline and where do you draw a line. 

The compiler here is Tailwind CLI and the framework is Tailwind CSS. Both are different projects and are 2 separate MIT licenses. You are trying to mix things up. I am talking about the framework which is in MIT and the attribution should be provided. Not for the build pipeline which you are talking about.

Also if you are pick on things with that, why would you put license for any JS library you use. They are just methods. As per your logic, there shouldn't be any "creative" work per se that is being done on it especially since at least in the frontend pipeline's perspective goes through and bundler/compiler like webpack and Vite.

> Anyway, this is getting very off topic. 

Agreed. My conversation was with respect to the comment mentioned from Tailwind CSS project. Not this ticket. Apologies for that.
Thanks for your POV. Appreciate your time.
